osquery/distributed/distributed.cpp

```diff
@@ -352,6 +399,88 @@ void Distributed::setCurrentRequestId(const std::string& cReqId) {
   currentRequestId_ = cReqId;
 }
 
+SQL Distributed::monitorNonnumeric(const std::string& name,
+                                   const std::string& query) {
+  // Snapshot the performance and times for the worker before running.
+  auto pid = std::to_string(PlatformProcess::getCurrentPid());
+  auto r0 = SQL::selectFrom({"resident_size", "user_time", "system_time"},
+                            "processes",
+                            "pid",
+                            EQUALS,
+                            pid);
+
+  using namespace std::chrono;
+  auto t0 = steady_clock::now();
+  SQL sql(query, true);
+
+  // Snapshot the performance after, and compare.
+  auto t1 = steady_clock::now();
+  auto r1 = SQL::selectFrom({"resident_size", "user_time", "system_time"},
+                            "processes",
+                            "pid",
+                            EQUALS,
+                            pid);
+  if (r0.size() > 0 && r1.size() > 0) {
+    // Always called while processes table is working.
+    uint64_t size = sql.rows().size();
+    recordQueryPerformance(
+        name, duration_cast<milliseconds>(t1 - t0).count(), size, r0[0], r1[0]);
+  }
+  return sql;
+}
+
+void Distributed::recordQueryPerformance(const std::string& name,
+                                         uint64_t delay_ms,
+                                         uint64_t size,
+                                         const Row& r0,
+                                         const Row& r1) {
+  RecursiveLock lock(performance_mutex_);
+  if (performance_.count(name) == 0) {
+    performance_[name] = QueryPerformance();
+  }
+
+  // Grab access to the non-const schedule item.
```

Comment:
Comment could use updating since this is no longer related to the schedule.

Reply:
I have no idea what this comment is even supposed to mean in the original code so I'll just delete it